Algorithm/BinaryTree/claude sonnet 4.6 extended/105. Construct Binary Tree from Preorder and Inorder Traversal/README.md

```
@@ -0,0 +1,759 @@
+
+<h2 id="tldr">アルゴリズム要点（TL;DR）</h2>
+
+> 💡 **TL;DR**（Too Long; Didn't Read）とは「長くて読めない人向けの要約」という意味です。ここでは「なんとなくこういう手順で解くんだな」というイメージを掴む章として読んでください。
+
+- **戦略**：「preorder の先頭 = ルート」という性質と「inorder のルート位置 = 左右の境界線」という性質を再帰的に利用して木を復元する
+- **前処理**：`inorder` の「値 → インデックス」を `dict`（ハッシュマップ）に登録しておく。なぜか：毎回 `list.index()` で線形探索すると全体が O(n²) になるが、`dict` なら O(1) で位置を取り出せる
+- **カーソル共有**：`preorder` を先頭から順に消費するカーソル変数を `nonlocal` で再帰関数間で共有する。なぜか：各再帰呼び出しが「今どのルートを処理するか」を順番通りに取り出す必要があるため
+- **時間計算量**：O(n)（各ノードをちょうど1回だけ処理）
+- **空間計算量**：O(n)（`dict` の n エントリ + 再帰スタックの高さ h 分）
+
+```
```

markdownlint-cli2 (0.22.0)

[warning] Fenced code blocks should have a language specified

(MD040, fenced-code-language)